Single Vxlan Device/Multiple Vlan Aware Bridges/Downstream VNI

sworleys opened the pull request:
This PR is a continuation of #8647 (merged) and #9204 (closed and reworked with additional DVNI commits)

The changes of this PR is with respect to three features Single Vxlan Device support, Multiple Vlan Aware Bridges, and Downstream VNI support in FRR.

Single Vxlan Device:
====================
Traditionally each VNI was mapped to a vxlan device (or interface) in kernel. This was posing challenges to VNI scale in kernel. For VNIs to scale better, kernel introduced single vxlan device construct where there is a single vxlan device created for the bridge and the vnis and its corresponding vlans and mcast-group are handled as map. SVD is analogous to Vlan Aware Bridges.

Sample Configuration:

    auto vxlan0
    iface vxlan0
          vxlan-local-tunnelip 27.0.0.9
          bridge-vids 1000-1001
          # vlan 1000-1001 maps to vni 1000-1001
          bridge-vlan-vni-map 1000-1001=1000-1001 <<<<<<<< VLAN-VNI map
          bridge-learning off
          vxlan-mcastgrp-map 1000=239.1.1.100 1001=239.1.1.200 <<<<<<<< VNI -> underlay mcast-group map

    auto bridge
    iface bridge
          bridge-vlan-aware yes
          bridge-ports vxlan0 swp1  <<<<<<<< Note the change here
          bridge-stp on
          bridge-vids 1000-1001
          bridge-pvid 1

As we can see from the above configuration, vlan-vni mapping is configured via bridge-vlan-vni-map attribute in /etc/network/interface and also, the mcast-group mapping for the vni is configured through vxlan-mcastgrp-map attribute. These two new attributes are introduced for supporting SVD.

Multiple Vlan Aware Bridges:
============================
Today, we only support single vlan aware bridge in FRR. This change also adds support for extending the bridge support in FRR to multiple vlan aware bridge (also called as MVAB). With MVAB, vlan is no more global and hence (vlan, bridge) identifies the L2 bridge domain.Each MVAB bridge hence can support overlapping vlans. However the assumption is the vnis across MVAB will not overlap. VNIs still continues to be global and should be uniquely configured across MVAB bridges.

Sample Configuration:

    auto vxlan0
    iface vxlan0
          vxlan-local-tunnelip 27.0.0.9
          bridge-vids 1000-1001
          bridge-vlan-vni-map 1000-1001=1000-1001 <<<<<<<< VLAN-VNI map
          bridge-learning off
          vxlan-mcastgrp-map 1000=239.1.1.100 1001=239.1.1.102 <<<<<<<< VNI -> underlay mcast-group map

    auto vxlan1
    iface vxlan1
          vxlan-local-tunnelip 27.0.0.9
          bridge-vids 1000-1001 1004
          bridge-vlan-vni-map 1000-1001=2000-2001 1004=2004 <<<<<<<< VLAN-VNI map
          bridge-learning off
          vxlan-mcastgrp-map 2000=239.1.1.103 2001=239.1.1.104 2004=239.1.1.105 <<<<<<<< VNI -> underlay mcast-group map

    auto bridge0
    iface bridge0
          bridge-vlan-aware yes
          bridge-ports vxlan0 swp1  <<<<<<<< Note the change here
          bridge-stp on
          bridge-vids 1000-1001
          bridge-pvid 1

    auto bridge1
    iface bridge1
          bridge-vlan-aware yes
          bridge-ports vxlan1 swp2  <<<<<<<< Note the change here
          bridge-stp on
          bridge-vids 1000-1001 1004
          bridge-pvid 1


Downstream VNI:
============================
Downstream VNI (symmetric EVPN route leaking) enables you to assign a VNI from a downstream remote VTEP through EVPN routes instead of configuring layer 3 VNIs globally across the network.

To configure a downstream VNI, you configure tenant VRFs as usual; however, to configure the desired route leaking, you define a route target import and, or export statement.

The route target import or export statement is in the format route-target import|export <asn>:<vni>; for example, route-target import 65101:6000. For route target import statements, you can use route-target import *:<vni> as an alternative. The asterisk (*) uses any ASN as a wildcard.

We make use of IP LWT encaps in the linux kernel to accomplish this.


```
router bgp 1 vrf vrf-red
 !
 address-family l2vpn evpn
  route-target import *:300
  route-target import auto
...
...
B>* 4.4.4.1/32 [20/0] via 2.2.2.2, vxlan0 (vrf default) onlink, label 300, weight 1, 01:30:22
...
...
4.4.4.1  encap ip id 300 src 0.0.0.0 dst 2.2.2.2 ttl 0 tos 0 via 2.2.2.2 dev vxlan0 proto bgp metric 20 onlink
```


---

New commands:

`show evpn access-vlan [IFNAME$if_name (1-4094)$vid | detail$detail] [json$json]`

c-po commented:
+1

riw777 commented on doc/user/zebra.rst:
```diff
@@ -580,6 +580,48 @@ nexthops are chosen to forward packets on.  Currently the Linux kernel
+Single Vxlan Device Support
+===========================
+
+FRR supports a new way of configuring VLAN-to-VNI mappings for EVPN-VXLAN,
```
I think I'd take out "new way" here ... at some point this is going to be the "old way," which might make the code confusing :-)

sworleys replied:
done

riw777 commented on bgpd/bgp_evpn.c:
```diff
@@ -7171,3 +7172,66 @@ void bgp_evpn_handle_resolve_overlay_index_unset(struct hash_bucket *bucket,
+ * Returns the label with the l3vni of the path's label stack.
+ *
+ * L3vni is always last label. Type5 will only
+ * have one label, Type2 will have two.
```
since we're only returning one l3vni here, I'm not certain why the comment here discusses two labels ... does this return the next to the last label in the case of a type2? 

just making certain I understand the logic here, rather than pointing out a problem in the code ...

sworleys replied:
you are correct. type2 has two labels and stores the l3vni as the second.

therefore grabbing the last will always be the l3vni for both route types

riw777 reviewed (approved):
looks good to me ... a couple of questions, nothing in the code ... it would really good to get someone who knows the zebra code better to look at those bits

cwmos commented:
I really like the functionality I believe is in this PR so thank you for all the effort!

Figuring out how to setup these things can be very difficult if you are not into the details of the implementation. So if https://docs.frrouting.org/en/latest/evpn.html could also be updated it would be super great! :-)

cwmos commented:
I have tested this PR and I believe I have identified an issue: It seems like when EVPN type 3 routes are installed in the VxLAN device, a VNI is not included.

In particular, when I run `bridge fdb show dev vxlan` in Linux I get:

```
00:00:aa:0a:03:01 vlan 10 extern_learn master br
00:00:aa:0a:03:02 vlan 10 extern_learn master br
00:00:aa:0a:02:01 vlan 10 extern_learn master br
00:00:aa:0a:02:02 vlan 10 extern_learn master br
5a:cb:88:2f:b3:a7 vlan 10 master br permanent
5a:cb:88:2f:b3:a7 vlan 1 master br permanent
5a:cb:88:2f:b3:a7 master br permanent
00:00:aa:0a:03:01 dst 192.168.1.3 src_vni 10 self extern_learn
00:00:aa:0a:03:02 dst 192.168.1.3 src_vni 10 self extern_learn
00:00:aa:0a:02:02 dst 192.168.1.2 src_vni 10 self extern_learn
00:00:00:00:00:00 dst 192.168.1.2 self permanent
00:00:00:00:00:00 dst 192.168.1.3 self permanent
00:00:aa:0a:02:01 dst 192.168.1.2 src_vni 10 self extern_learn
```

Notice that the two routes for 00:00:00:00:00:00 do not have a src_vni.

Further, when the VxLAN device receives an ARP request locally, it does not forward it on the network.

However, if I run:

```
bridge fdb delete dev vxlan 00:00:00:00:00:00 dst 192.168.1.2
bridge fdb delete dev vxlan 00:00:00:00:00:00 dst 192.168.1.3
bridge fdb append dev vxlan 00:00:00:00:00:00 dst 192.168.1.2 src_vni 10
bridge fdb append dev vxlan 00:00:00:00:00:00 dst 192.168.1.3 src_vni 10 
```

Then `bridge fdb show dev vxlan` gives:

```
00:00:aa:0a:03:01 vlan 10 extern_learn master br
00:00:aa:0a:03:02 vlan 10 extern_learn master br
00:00:aa:0a:02:01 vlan 10 extern_learn master br
00:00:aa:0a:02:02 vlan 10 extern_learn master br
5a:cb:88:2f:b3:a7 vlan 10 master br permanent
5a:cb:88:2f:b3:a7 vlan 1 master br permanent
5a:cb:88:2f:b3:a7 master br permanent
00:00:00:00:00:00 dst 192.168.1.2 src_vni 10 self permanent
00:00:00:00:00:00 dst 192.168.1.3 src_vni 10 self permanent
00:00:aa:0a:03:01 dst 192.168.1.3 src_vni 10 self extern_learn
00:00:aa:0a:03:02 dst 192.168.1.3 src_vni 10 self extern_learn
00:00:aa:0a:02:02 dst 192.168.1.2 src_vni 10 self extern_learn
00:00:aa:0a:02:01 dst 192.168.1.2 src_vni 10 self extern_learn
```

and the VxLAN device forwards ARP requets as expected.

riw777 commented:
> I have tested this PR and I believe I have identified an issue: It seems like when EVPN type 3 routes are installed in the VxLAN device, a VNI is not included.

Should this case be covered in the topo tests?

siddanmk commented:
@sworleys and @donaldsharp, can we prioritize this?

riw777 commented:
lint is all in `#define` stuff ... waiting on the conflict, nits, etc.

sworleys commented:
Fixed the type3 FLOOD issue mentioned above and added a topotest for it.

Also added a few additional fixes we had internally that needed upstreaming.

donaldsharp commented:
ci:rerun

riw777 commented:
https://github.com/FRRouting/frr/pull/12364/commits/5e1f53471d4689dc8fef5cee4c778101730c68b9
https://github.com/FRRouting/frr/pull/12364/commits/3434f0bb26aad1a96c46d599292c11b335f558c9

Are missing "signed-off-by" ... can we get these fixed so I can merge this?

Thanks!

sworleys commented:
ci:rerun

sworleys commented:
ci:rerun

sworleys commented:
Added a commit to guard the presence of `bridge fdb get` on systems where the iproute2/kernel version don't match...

see [here](https://frrouting.slack.com/archives/C016LQGFWLQ/p1676268218955959) for context


and rebased.
